plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftPageSourceProvider.java

```diff
@@ -0,0 +1,142 @@
+import org.apache.parquet.io.MessageColumnIO;
+import org.apache.parquet.schema.MessageType;
+import org.joda.time.DateTimeZone;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.memory.context.AggregatedMemoryContext.newSimpleAggregatedMemoryContext;
+import static io.trino.parquet.ParquetTypeUtils.constructField;
+import static io.trino.parquet.ParquetTypeUtils.getColumnIO;
+import static io.trino.parquet.ParquetTypeUtils.getDescriptors;
+import static io.trino.parquet.ParquetTypeUtils.lookupColumnByName;
+import static io.trino.parquet.metadata.PrunedBlockMetadata.createPrunedColumnsMetadata;
+import static io.trino.plugin.redshift.RedshiftErrorCode.REDSHIFT_PARQUET_CURSOR_ERROR;
+import static io.trino.plugin.redshift.RedshiftPageSource.handleException;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class RedshiftPageSourceProvider
+        implements ConnectorPageSourceProvider
+{
+    private final ConnectorRecordSetProvider recordSetProvider;
+    private final TrinoFileSystemFactory fileSystemFactory;
+
+    public RedshiftPageSourceProvider(ConnectorRecordSetProvider recordSetProvider, TrinoFileSystemFactory fileSystemFactory)
+    {
+        this.recordSetProvider = requireNonNull(recordSetProvider, "recordSetProvider is null");
+        this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
+    }
+
+    @Override
+    public ConnectorPageSource createPageSource(
+            ConnectorTransactionHandle transaction,
+            ConnectorSession session,
+            ConnectorSplit split,
+            ConnectorTableHandle table,
+            List<ColumnHandle> columns,
+            DynamicFilter dynamicFilter)
+    {
+        if (split instanceof JdbcSplit) {
+            return new RecordPageSource(recordSetProvider.getRecordSet(transaction, session, split, table, columns));
+        }
+
+        String path = ((RedshiftUnloadSplit) split).path();
+        Location location = Location.of(path);
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        TrinoInputFile inputFile = fileSystem.newInputFile(location);
+        ParquetReader parquetReader;
+        try {
+            parquetReader = parquetReader(inputFile, columns);
+        }
+        catch (IOException e) {
+            throw new TrinoException(REDSHIFT_PARQUET_CURSOR_ERROR, format("Failed to open Parquet file: %s", path), e);
+        }
+        return new RedshiftPageSource(parquetReader, () -> fileSystemFactory.create(session).deleteFile(location));
+    }
+
+    private ParquetReader parquetReader(TrinoInputFile inputFile, List<ColumnHandle> columns)
+            throws IOException
+    {
+        ParquetReaderOptions options = new ParquetReaderOptions();
+        TrinoParquetDataSource dataSource = new TrinoParquetDataSource(inputFile, options);
```

Comment:
FileFormatDataSourceStats should be injected and exported in jmx similar to other connectors reading from parquet